platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ResolutionHost.java

```diff
@@ -42,13 +50,71 @@ default DisplayName displayName(String type) {
      */
-    default void rethrowFailure(String type, Collection<Throwable> failures) {
-        mapFailure(type, failures).ifPresent(e -> {
+    InternalProblems getProblems();
+
+    /**
+     * Rethrows the provided failures, doing nothing if the list of failures is empty.
+     * <p>
+     * If any of the failures (or their ancestor causes) are {@link ReportableAsProblem}, they will all be reported to the problems
+     * service available on this type via {@link #getProblems()}.
+     *
+     * @param resolutionType what was resolved, e.g. "dependencies", "artifacts", "files"
+     * @param failures the exceptions encountered during resolution
+     */
+    default void rethrowFailuresAndReportProblems(String resolutionType, Collection<Throwable> failures) {
+        reportProblems(failures);
+        consolidateFailures(resolutionType, failures).ifPresent(e -> {
             throw e;
         });
     }
 
-    Optional<? extends ResolveException> mapFailure(String type, Collection<Throwable> failures);
+    /**
+     * Consolidates the given failures into a single exception, if there are any.
+     *
+     * @param resolutionType what was resolved, e.g. "dependencies", "artifacts", "files"
+     * @param failures the exceptions encountered during resolution
+     * @return an {@link org.gradle.internal.exceptions.MultiCauseException} containing all the failures, or {@link Optional#empty()} if there are no failures
+     */
+    Optional<? extends ResolveException> consolidateFailures(String resolutionType, Collection<Throwable> failures);
+
+    /**
+     * If the given failure (or their ancestor causes) are {@link ReportableAsProblem}, they will all be reported to the problems
+     * service available on this type via {@link #getProblems()}.
+     *
+     * @param failure the exception to inspect
+     */
+    default void reportProblems(Throwable failure) {
+        reportProblems(Collections.singleton(failure));
+    }
+
+    /**
+     * If the given failures (or their ancestor causes) are {@link ReportableAsProblem}, they will all be reported to the problems
+     * service available on this type via {@link #getProblems()}.
+     *
+     * @param failures the exceptions to inspect
+     */
+    default void reportProblems(Collection<Throwable> failures) {
+        Queue<Throwable> exceptionQueue = new LinkedList<>(failures);
+        Throwable current = exceptionQueue.poll();
+
+        while (current != null) {
+            if (current instanceof ReportableAsProblem) {
+                ((ReportableAsProblem<?>) current).reportAsProblem(getProblems());
+            }
+
+            if (current instanceof MultiCauseException) {
+                exceptionQueue.addAll(((MultiCauseException) current).getCauses());
+            } else {
+                if (current.getCause() != null) {
+                    exceptionQueue.add(current.getCause());
+                }
+            }
+
+            current = exceptionQueue.poll();
```

Comment:
💅 I think you can avoid having two calls to `exceptionQueue.poll` if you put the `poll` at the top of the loop:
```
while (!exceptionQueue.isEmpty()) {
    Throwable current = exceptionQueue.poll();
    // ...
}
```

Reply:
Example of a similar traversal: https://github.com/gradle/gradle/blob/b09c974bf1f9341a781f0b4abb61834db61fedd5/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java#L1176C18-L1191

Reply:
OK, I'll update this.